File: kinv/report.py
"""The whole reading of a board, in the shape the page and every command consume."""

from __future__ import annotations

import os
import threading
from typing import Callable

from kinv.adapters.kicad.bom import read_bom_csv
from kinv.adapters.kicad.cli import export_bom
from kinv.adapters.kicad.hierarchy import project_sheets_or_directory
from kinv.adapters.kicad.libraries import library_table, resolve_footprint
from kinv.adapters.kicad.mod import read_footprint_file
from kinv.adapters.store.inventory import inventory_signature, read_assignments, read_catalog
from kinv.adapters.store.project_state import buy_signature, read_buy_choices, read_solved, solved_signature
from kinv.core.consolidate.findings import compare_columns, consolidate, parts_by_key
from kinv.core.consolidate.identity import finding_id, finding_signature
from kinv.core.footprint.audit import audit_footprints
from kinv.core.footprint.check import check_footprint
from kinv.core.footprint.measure import measure
from kinv.core.footprint.name import parse_footprint_name
from kinv.core.inventory.package_check import package_issues
from kinv.core.inventory.resolve import now_iso, resolve_parts
from kinv.core.parse.spec import analyze_bom, apply_buy_choices, apply_mountability
from kinv.js import js_sorted, num_str


def source_signature(project: str) -> str:
    """A cheap fingerprint of the project's schematics, so kicad-cli only runs when something changed.

    Modification time *and* total size: a hierarchy save writes every sheet at
    once, and on a fast disk several can share a timestamp.
    """
    try:
        if os.path.splitext(project)[1].lower() == ".csv":
            stat = os.stat(project)
            return f"{num_str(stat.st_mtime_ns / 1e6)}:{stat.st_size}"
        directory = os.path.dirname(project) or "."
        if not os.path.exists(directory):
            return ""

        # Only the sheets the design is made of. KiCad's `_autosave-*` files
        # change on a timer and would re-export the BOM for an edit nobody made.
        sheets = project_sheets_or_directory(
            project,
            [os.path.join(directory, name) for name in os.listdir(directory) if name.endswith(".kicad_sch")],
        )
        newest = 0.0
        size = 0
        files = 0
        for path in sheets:
            stat = os.stat(path)
            newest = max(newest, stat.st_mtime_ns / 1e6)
            size += stat.st_size
            files += 1
        return "" if files == 0 else f"{num_str(newest)}:{size}:{files}"
    except OSError:
        return ""


def load_project(project: str, group_by: str | None = None, exclude_dnp: bool = False, include_excluded: bool = False) -> list[dict]:
    """A ``.csv`` is read directly; anything else goes through kicad-cli."""
    if project.lower().endswith(".csv"):
        lines = read_bom_csv(project)
    else:
        lines = export_bom(project, group_by=group_by, exclude_dnp=exclude_dnp)
    return analyze_bom(lines, [] if include_excluded else None)


def analyze_project(
    project: str,
    group_by: str | None = None,
    exclude_dnp: bool = False,
    include_excluded: bool = False,
) -> tuple[list[dict], list[dict]]:
    """Reads the board the way every command must read it: one reading, one answer."""
    raw = load_project(project, group_by=group_by, exclude_dnp=exclude_dnp, include_excluded=include_excluded)

    # Footprint libraries are read straight from disk; a library that has gone
    # missing must degrade to "cannot verify", never take the report down.
    try:
        table = library_table(project)

        def pads(reference: str):
            path = resolve_footprint(reference, table)
            return None if path is None else read_footprint_file(path)["pads"]

        footprints = audit_footprints(raw, pads)
    except Exception:  # noqa: BLE001 — any failure here means "cannot verify"
        footprints = []

    # Geometry overrules the class-and-part-number guess in both directions;
    # your own decisions overrule both. `--include-excluded` sets all aside.
    mountability = {f["reference"]: f["mountable"] for f in footprints if not f["unresolved"]}
    choices = {c["key"]: c["buy"] for c in read_buy_choices(project)}
    if include_excluded:
        lines = raw
    else:
        lines = apply_buy_choices(apply_mountability(raw, mountability.get), choices.get)
    return lines, footprints


def build_report(
    project: str,
    group_by: str | None = None,
    exclude_dnp: bool = False,
    near_value_percent: float | None = None,
    include_excluded: bool = False,
) -> dict:
    signature = source_signature(project)
    lines, footprints = analyze_project(project, group_by, exclude_dnp, include_excluded)

    found = consolidate(lines, near_value_percent)

    # A mark holds only while the finding still says what it said when it was settled.
    marks = {m["id"]: m for m in read_solved(project)}
    identified = [{**f, "id": finding_id(f), "signature": finding_signature(f)} for f in found]
    settled = [f for f in identified if marks.get(f["id"], {}).get("signature") == f["signature"]]
    settled_ids = {f["id"] for f in settled}
    findings = [f for f in identified if f["id"] not in settled_ids]

    parts = parts_by_key(lines)
    resolutions = resolve_parts(parts, read_catalog(), read_assignments())
    issues = [*(issue for line in lines for issue in line["issues"]), *package_issues(parts, resolutions)]

    bought = [line for line in lines if line["excluded"] is None]
    left_out = [line for line in lines if line["excluded"] is not None]
    columns: dict[str, None] = {}
    for part in parts:
        for name in part["fields"]:
            columns[name] = None

    def with_solved_at(f: dict) -> dict:
        at = marks.get(f["id"], {}).get("at")
        return f if at is None else {**f, "solvedAt": at}

    return {
        "project": project,
        "generatedAt": now_iso(),
        "sourceSignature": signature,
        "summary": {
            # Placements and lines count what will be bought; the excluded ones
            # are reported separately rather than silently folded in.
            "placements": sum(len(line["line"]["refs"]) for line in bought),
            "lines": len(bought),
            "parts": len(parts),
        },
        "counts": {
            "errors": len([i for i in issues if i["severity"] == "error"]),
            "warnings": len([i for i in issues if i["severity"] == "warning"]),
            "infos": len([i for i in issues if i["severity"] == "info"]),
            "findings": len(findings),
            "settled": len(settled),
            "assigned": len([r for r in resolutions if r["assignment"] is not None]),
            "resolved": len([p for p in parts if p["resolved"]]),
            "excluded": sum(len(line["line"]["refs"]) for line in left_out),
            "footprintFindings": sum(len(f["findings"]) for f in footprints),
        },
        "findings": findings,
        "settled": [with_solved_at(f) for f in settled],
        "issues": issues,
        "parts": parts,
        "partColumns": js_sorted(columns, compare_columns),
        "resolutions": resolutions,
        "footprints": footprints,
        "excluded": [
            {
                "key": line["key"],
                "value": line["line"]["value"],
                "footprint": line["line"]["footprint"],
                "reason": line["excluded"],
                "refs": line["line"]["refs"],
            }
            for line in left_out
        ],
    }


def create_report_cache(project: str, **options) -> Callable[[], dict]:
    """Rebuilds only when the schematics, the catalog or a decision have changed.

    Thread-safe: the server answers requests on several threads, and two polls
    arriving together must share one kicad-cli run rather than start two.
    """
    lock = threading.Lock()
    state: dict = {"report": None, "signature": None}

    def get() -> dict:
        # Stamped with the signature as it was *before* the build: a save that
        # lands while the build runs is still a change next time.
        signature = f"{source_signature(project)}|{solved_signature(project)}|{buy_signature(project)}|{inventory_signature()}"
        with lock:
            if state["report"] is not None and state["signature"] == signature:
                return state["report"]
            report = build_report(project, **options)
            state["report"] = report
            state["signature"] = signature
            return report

    return get


def footprint_detail(project: str, reference: str) -> dict | None:
    """Everything the browser needs to draw one footprint and measure on it."""
    path = resolve_footprint(reference, library_table(project))
    if path is None:
        return None
    file = read_footprint_file(path)
    measured = measure(file["pads"])
    declared = parse_footprint_name(reference)
    return {
        "reference": reference,
        "name": file["name"],
        "path": path,
        "pads": file["pads"],
        "courtyard": file["courtyard"],
        "measured": measured,
        "declared": declared,
        "findings": check_footprint(declared, measured),
    }

File: src/report.ts
import { existsSync, readdirSync, statSync } from "node:fs";
import { dirname, extname, join } from "node:path";

import { projectSheetsOrDirectory } from "./adapters/kicad/hierarchy.js";
import { libraryTable, resolveFootprint } from "./adapters/kicad/libraries.js";
import { inventorySignature, readAssignments, readCatalog } from "./adapters/store/inventory.js";
import { buySignature, readBuyChoices } from "./adapters/store/buy.js";
import { readSolved, solvedSignature } from "./adapters/store/solved.js";
import { readFootprintFile } from "./adapters/kicad/mod.js";
import {
  compareColumns,
  consolidate,
  partsByKey,
  type Finding,
  type PartRollup,
} from "./core/consolidate/findings.js";
import { findingId, findingSignature } from "./core/consolidate/identity.js";
import { auditFootprints, type FootprintAudit } from "./core/footprint/audit.js";
import { packageIssues } from "./core/inventory/package-check.js";
import { resolveParts } from "./core/inventory/resolve.js";
import type { Resolution } from "./core/inventory/types.js";
import { checkFootprint, type FootprintFinding } from "./core/footprint/check.js";
import { measure, type MeasuredPad, type Measurement } from "./core/footprint/measure.js";
import { parseFootprintName, type DeclaredFootprint } from "./core/footprint/name.js";
import { applyBuyChoices, applyMountability } from "./core/parse/spec.js";
import type { AnalyzedLine, Issue } from "./core/types.js";
import { loadProject } from "./cli/commands/check.js";

export type Report = {
  project: string;
  generatedAt: string;
  /** Fingerprint of the sources this was built from; the cache key. */
  sourceSignature: string;
  summary: { placements: number; lines: number; parts: number };
  counts: {
    errors: number;
    warnings: number;
    infos: number;
    findings: number;
    resolved: number;
    footprintFindings: number;
    excluded: number;
    settled: number;
    /** Parts with a catalog assignment: step 2's progress bar. */
    assigned: number;
  };
  /** Findings still open, each carrying the name a "solved" mark is filed under. */
  findings: OpenFinding[];
  /** Findings you have settled, kept so they can be reopened. */
  settled: OpenFinding[];
  issues: Issue[];
  parts: PartRollup[];
  /** What each part is bought as, so far. One entry per part, same order. */
  resolutions: Resolution[];
  /**
   * The BOM columns worth a column in the parts table: every field at least one
   * part actually carries. A field the export asked for and nobody filled in
   * would otherwise widen the table by a column of blanks.
   */
  partColumns: string[];
  footprints: FootprintAudit[];
  excluded: { key: string; value: string; footprint: string; reason: string; refs: string[] }[];
};

/**
 * A cheap fingerprint of the project's schematics, so kicad-cli only runs when
 * something actually changed.
 *
 * Modification time *and* total size, because a filesystem's mtime can be no
 * finer than a millisecond: saving a hierarchy writes every sheet at once, and
 * on a fast disk several of them can share a timestamp. Size catches the edit
 * that a timestamp alone would miss.
 */
export function sourceSignature(project: string): string {
  try {
    if (extname(project).toLowerCase() === ".csv") {
      const stat = statSync(project);
      return `${stat.mtimeMs}:${stat.size}`;
    }
    const dir = dirname(project);
    if (!existsSync(dir)) return "";

    // Only the sheets the design is made of. KiCad's own `_autosave-*.kicad_sch`
    // sits in the same directory and changes on a timer, which would re-export
    // the BOM for an edit nobody made.
    const sheets = projectSheetsOrDirectory(
      project,
      readdirSync(dir)
        .filter((name) => name.endsWith(".kicad_sch"))
        .map((name) => join(dir, name)),
    );

    let newest = 0;
    let bytes = 0;
    let files = 0;
    for (const path of sheets) {
      const stat = statSync(path);
      newest = Math.max(newest, stat.mtimeMs);
      bytes += stat.size;
      files += 1;
    }
    return files === 0 ? "" : `${newest}:${bytes}:${files}`;
  } catch {
    return "";
  }
}

/** A finding plus the identity the solved list files it under. */
export type OpenFinding = Finding & { id: string; signature: string; solvedAt?: string };

export type ReportOptions = {
  groupBy?: string;
  excludeDnp?: boolean;
  nearValuePercent?: number;
  includeExcluded?: boolean;
};

/**
 * Reads the board the way every command must read it.
 *
 * `kinv check` used to do its own `loadProject` and skip the geometry step
 * below, so it reported 81 parts where the UI reported 83 — the two Würth SMD
 * standoffs, which the class rule throws out as mounting holes and the copper
 * puts back. One reading, one answer.
 */
export async function analyzeProject(
  project: string,
  options: ReportOptions = {},
): Promise<{ lines: AnalyzedLine[]; footprints: FootprintAudit[] }> {
  const rawLines = await loadProject(project, {
    ...(options.groupBy !== undefined ? { groupBy: options.groupBy } : {}),
    ...(options.excludeDnp === true ? { excludeDnp: true } : {}),
    ...(options.includeExcluded === true ? { includeExcluded: true } : {}),
  });

  // Footprint libraries are read straight from disk; a library that has gone
  // missing must degrade to "cannot verify", never take the whole report down.
  let footprints: FootprintAudit[] = [];
  try {
    const table = libraryTable(project);
    footprints = auditFootprints(rawLines, (reference) => {
      const path = resolveFootprint(reference, table);
      return path === undefined ? undefined : readFootprintFile(path).pads;
    });
  } catch {
    footprints = [];
  }

  // Measure where you can, guess by name only where you cannot.
  //
  // Geometry is authoritative: a footprint with neither solder paste nor a
  // plated hole cannot be soldered to, and one that has either can. That
  // overrides the class-and-part-number guess in *both* directions — it drops
  // test pads no naming rule would catch, and it keeps a Würth SMD standoff
  // that the naming rule would have thrown away for want of an MPN. The guess
  // survives only for footprints that could not be found in any library.
  const mountability = new Map(
    footprints.filter((f) => !f.unresolved).map((f) => [f.reference, f.mountable]),
  );
  // And your own decisions last of all: the rules describe the usual board,
  // you are looking at this one. `--include-excluded` asks for everything the
  // board has, rules and decisions alike set aside.
  const choices = new Map(readBuyChoices(project).map((c) => [c.key, c.buy]));
  const lines =
    options.includeExcluded === true
      ? rawLines
      : applyBuyChoices(
          applyMountability(rawLines, (footprint) => mountability.get(footprint)),
          (key) => choices.get(key),
        );

  return { lines, footprints };
}

export async function buildReport(project: string, options: ReportOptions = {}): Promise<Report> {
  const signature = sourceSignature(project);
  const { lines, footprints } = await analyzeProject(project, options);

  const found = consolidate(lines, {
    ...(options.nearValuePercent !== undefined
      ? { nearValuePercent: options.nearValuePercent }
      : {}),
  });

  // A mark holds only while the finding still says what it said when it was
  // settled: "10k in 0402 and 0603" stays quiet, and speaks up again the day a
  // third package joins it.
  const marks = new Map(readSolved(project).map((m) => [m.id, m]));
  const identified: OpenFinding[] = found.map((f) => ({
    ...f,
    id: findingId(f),
    signature: findingSignature(f),
  }));
  const settled = identified.filter((f) => marks.get(f.id)?.signature === f.signature);
  const settledIds = new Set(settled.map((f) => f.id));
  const findings = identified.filter((f) => !settledIds.has(f.id));
  // The one check that needs the catalog: the schematic cannot say that you
  // went and bought 0603 resistors for an 0402 land pattern.
  const parts = partsByKey(lines);
  const resolutions = resolveParts(parts, readCatalog(), readAssignments());
  const issues = [...lines.flatMap((l) => l.issues), ...packageIssues(parts, resolutions)];

  return {
    project,
    generatedAt: new Date().toISOString(),
    sourceSignature: signature,
    summary: {
      // Placements and lines count what will be bought; the excluded ones are
      // reported separately rather than silently folded into the totals.
      placements: lines
        .filter((l) => l.excluded === undefined)
        .reduce((n, l) => n + l.line.refs.length, 0),
      lines: lines.filter((l) => l.excluded === undefined).length,
      parts: parts.length,
    },
    counts: {
      errors: issues.filter((i) => i.severity === "error").length,
      warnings: issues.filter((i) => i.severity === "warning").length,
      infos: issues.filter((i) => i.severity === "info").length,
      findings: findings.length,
      settled: settled.length,
      assigned: resolutions.filter((r) => r.assignment !== undefined).length,
      resolved: parts.filter((p) => p.resolved).length,
      excluded: lines
        .filter((l) => l.excluded !== undefined)
        .reduce((n, l) => n + l.line.refs.length, 0),
      footprintFindings: footprints.reduce((n, f) => n + f.findings.length, 0),
    },
    findings,
    settled: settled.map((f) => {
      const at = marks.get(f.id)?.at;
      return at === undefined ? f : { ...f, solvedAt: at };
    }),
    issues,
    parts,
    partColumns: [...new Set(parts.flatMap((p) => Object.keys(p.fields)))].sort(compareColumns),
    resolutions,
    footprints,
    excluded: lines
      .filter((l) => l.excluded !== undefined)
      .map((l) => ({
        key: l.key,
        value: l.line.value,
        footprint: l.line.footprint,
        reason: l.excluded as string,
        refs: l.line.refs,
      })),
  };
}

/** Rebuilds only when the schematics have changed since the cached run. */
export function createReportCache(project: string, options: ReportOptions = {}) {
  let cached: Report | undefined;
  /**
   * The composite the cached report was built from — kept here rather than
   * read back off `cached.sourceSignature`, which is the schematic
   * fingerprint alone and so never equalled the composite. The comparison was
   * therefore always false: every poll rebuilt the report, re-ran kicad-cli,
   * and handed the page a new `generatedAt` — which re-rendered the view and
   * took whatever was half-typed in a box with it.
   */
  let cachedSignature: string | undefined;
  let inFlight: Promise<Report> | undefined;

  return async function get(): Promise<Report> {
    // Assigning a part, settling a finding or marking one "do not buy"
    // touches no schematic; without these the page would keep serving a report
    // that predates the decision.
    const signature =
      `${sourceSignature(project)}|${solvedSignature(project)}|` +
      `${buySignature(project)}|${inventorySignature()}`;
    if (cached && cachedSignature === signature) return cached;
    if (inFlight) return inFlight;

    // Stamped with the signature as it was *before* the build: a save that
    // lands while the build is running is then still a change next time,
    // rather than one this report gets the credit for having read.
    inFlight = buildReport(project, options)
      .then((report) => {
        cached = report;
        cachedSignature = signature;
        return report;
      })
      .finally(() => {
        inFlight = undefined;
      });
    return inFlight;
  };
}

/** Everything the browser needs to draw one footprint and measure on it. */
export type FootprintDetail = {
  reference: string;
  name: string;
  path: string;
  pads: MeasuredPad[];
  courtyard: { x: number; y: number; width: number; height: number } | undefined;
  measured: Measurement;
  declared: DeclaredFootprint;
  findings: FootprintFinding[];
};

export function footprintDetail(project: string, reference: string): FootprintDetail | undefined {
  const path = resolveFootprint(reference, libraryTable(project));
  if (path === undefined) return undefined;

  const file = readFootprintFile(path);
  const measured = measure(file.pads);
  const declared = parseFootprintName(reference);

  return {
    reference,
    name: file.name,
    path,
    pads: file.pads,
    courtyard: file.courtyard,
    measured,
    declared,
    findings: checkFootprint(declared, measured),
  };
}
